Docs: Changes for coordinator improvements (#14590)
apache/druid: docs/configuration/index.md

|`millisToWaitBeforeDeleting`|How long does the Coordinator need to be a leader before it can start marking overshadowed segments as unused in metadata storage.|900000 (15 mins)|
|`mergeBytesLimit`|The maximum total uncompressed size in bytes of segments to merge.|524288000L|
|`mergeSegmentsLimit`|The maximum number of segments that can be in a single [append task](../ingestion/tasks.md).|100|

> |`smartSegmentLoading`|Whether to turn on the new ["smart"-mode of segment loading](#smart-segment-loading) which dynamically computes the optimal values of several parameters that maximize Coordinator performance.|true|

Reviewer: ```suggestion
|`smartSegmentLoading`|Enables ["smart" segment loading mode](#smart-segment-loading) which dynamically computes the optimal values of several properties that maximize Coordinator performance.|true|
```

|`maxSegmentsToMove`|The maximum number of segments that can be moved at any given time.|100|
|`replicantLifetime`|The maximum number of Coordinator runs for which a segment can wait in the load queue of a Historical before Druid raises an alert.|15|

> |`replicationThrottleLimit`|The maximum number of segment replicas that can be assigned to a historical tier in a single Coordinator run. This property prevents historicals from becoming overwhelmed when loading extra replicas of segments that are already available in the cluster.|500|

Reviewer 2: (Nit): The maximum number of non-primary segment replicas that can be assigned to a historical tier in a single Coordinator run. This property prevents historicals from becoming overwhelmed when loading extra replicas of segments that are already available in the tier.

Author: We are trying to get rid of the primary / non-primary terminology, since replica already means a copy. We will be completely removing the other config `maxNonPrimaryReplicantsToLoad` soon.

> |`replicationThrottleLimit`|The maximum number of segment replicas that can be assigned to a Historical tier in a single Coordinator run. This parameter is a defensive measure to prevent Historicals from getting overwhelmed loading extra replicas of segments that are already available in the cluster.|500|

Reviewer: ```suggestion
|`replicationThrottleLimit`|The maximum number of segment replicas that can be assigned to a historical tier in a single Coordinator run. This property prevents historicals from becoming overwhelmed when loading extra replicas of segments that are already available in the cluster.|500|
```

> |`balancerComputeThreads`|Thread pool size for computing moving cost of segments during segment balancing. Consider increasing this if you have a lot of segments and moving segments starts to get stuck.|1|

Reviewer: ```suggestion
|`balancerComputeThreads`|Thread pool size for computing moving cost of segments during segment balancing. Consider increasing this if you have a lot of segments and moving segments begins to stall.|1|
```

|`maxNonPrimaryReplicantsToLoad`|The maximum number of replicas that can be assigned across all tiers in a single Coordinator run. This parameter serves the same purpose as `replicationThrottleLimit` except this limit applies at the cluster-level instead of per tier. The default value essentially means that there is no limit on the number of replicas assigned per coordination cycle. If you want to use a non-default value for this config, you may want to start with it being `~20%` of the number of segments found on the Historical server with the most segments. Use the Druid metric, `coordinator/time` with the filter `duty=org.apache.druid.server.coordinator.duty.RunRules` to see how different values of this config impact your Coordinator execution time.|`Integer.MAX_VALUE` (i.e. no limit)|

##### Smart segment loading

The `smartSegmentLoading` mode of the Coordinator makes configuring it for segment loading and balancing much easier.
In this mode, the Coordinator does not require the user to provide values of the following parameters and computes them automatically instead.

> If provided, the values are simply ignored.

Reviewer 3: should be put in bold.

Author: Done.

The computed values are based on the current state of the cluster and are meant to optimize the performance of the Coordinator.

> |Property|Computed value|Explanation|

Reviewer: ```suggestion
|Property|Computed value|Description|
```

|--------|--------------|-----------|
|`useRoundRobinSegmentAssignment`|true|Speeds up segment assignment|

> |`maxSegmentsInNodeLoadingQueue`|0|Removes the limit on load queue size|

Reviewer: ```suggestion
|`maxSegmentsInNodeLoadingQueue`|0|Removes the limit on load queue size.|
```

> |`replicationThrottleLimit`|2% of used segments, minimum value = 100|Ensures that replication is not done too aggressively in case of a historical disappearing only intermittently.|

Reviewer: ```suggestion
|`replicationThrottleLimit`|2% of used segments, minimum value 100|Prevents aggressive replication when a historical disappears only intermittently.|
```

> |`replicantLifetime`|60|Allows segments to wait about an hour (assuming a coordinator period of 1 minute) in the load queue before an alert is raised. This value is higher than the previous default of 15 because in `smartSegmentLoading` mode, load queues are not limited by size. Thus, segments might get assigned to a load queue even if the corresponding server is slow to load them.|

Reviewer: ```suggestion
|`replicantLifetime`|60|Allows segments to wait about an hour (assuming a Coordinator period of 1 minute) in the load queue before an alert is raised. In `smartSegmentLoading` mode, load queues are not limited by size. Segments might therefore assigned to a load queue even if the corresponding server is slow to load them.|
```

|`maxNonPrimaryReplicantsToLoad`|`Integer.MAX_VALUE` (no limit)|This throttling is already handled by `replicationThrottleLimit`.|

> |`maxSegmentsToMove`|2% of used segments, minimum value = 100, maximum value = 1000|Ensures that some segments are always moving in the cluster to keep it well balanced. The maximum value keeps the coordinator run times bounded.|

Reviewer: ```suggestion
|`maxSegmentsToMove`|2% of used segments, minimum value 100, maximum value 1000|Ensures that some segments are always moving in the cluster to keep it well balanced. The maximum value keeps the Coordinator run times bounded.|
```

> |`decommissioningMaxPercentOfMaxSegmentsToMove`|100|Prioritizes move of segments from decommissioning servers so that they can be terminated quickly.|

Reviewer: ```suggestion
|`decommissioningMaxPercentOfMaxSegmentsToMove`|100|Prioritizes the move of segments from decommissioning servers so that they can be terminated quickly.|
```

> When `smartSegmentLoading` is disabled, the configured values of these parameters are used without any modification.

Reviewer: ```suggestion
When `smartSegmentLoading` is disabled, Druid uses the configured values of these properties.
```

> You should disable this mode only if you want to explicitly set the value of any of the above parameters.

Reviewer: ```suggestion
Disable `smartSegmentLoading` only if you want to explicitly set the values of any of the above properties.
```
